### PR TITLE
[NP-5360] Cron to cleanup tempFileDAO

### PR DESCRIPTION
**Pull request by Mykola97**
https://nanopay.atlassian.net/browse/NP-5360

## Dependency
- [ ] https://github.com/kgrgreer/foam3/pull/823

### CODE REVIEW

File: src/cronjobs [hunk 1]
```diff
@@ -116,3 +116,47 @@ p({
...
+    import java.util.Calendar;
+    import java.util.Date;
+    import java.util.List;
+
+    DAO tempFileDAO = (DAO) x.get("tempFileDAO");
+
+    Calendar calendar = Calendar.getInstance();
+    calendar.setTime(new Date());
+    calendar.add(Calendar.HOUR_OF_DAY, -4);
+    Date forHourBefore = calendar.getTime();
```
Comment (jlhughes) [inline src/cronjobs]:
```suggestion
    Date threshold = calendar.getTime();
```

File: src/cronjobs [hunk 2]
```diff
@@ -116,3 +116,47 @@ p({
...
+
+    List tempFiles = ((ArraySink) tempFileDAO.select(new ArraySink())).getArray();
```
Comment (jlhughes) [inline src/cronjobs]: Additionally this could simply be 
where(LT(CREATED, threshold).select().removeAll();

File: src/cronjobs [hunk 3]
```diff
@@ -116,3 +116,47 @@ p({
...
+  "description":"Cron to clean up files which exist more than 4 hours",
+  schedule:{
+    class:"foam.nanos.cron.IntervalSchedule",
+    duration:{
+      class:"foam.nanos.cron.TimeHMS",
+      hour:4,
+      minute:0,
+      seconds:0
+    }
+  },
```
Comment (lchanmann) [inline src/cronjobs]: Could run every hour instead and setup the TTL constant in the code.

File: src/cronjobs [hunk 4]
```diff
@@ -116,3 +116,47 @@ p({
...
+    import foam.nanos.fs.TempFile;
+    import java.io.File;
+    import java.util.Calendar;
+    import java.util.Date;
+    import java.util.List;
+
+    DAO tempFileDAO = (DAO) x.get("tempFileDAO");
+
+    Calendar calendar = Calendar.getInstance();
+    calendar.setTime(new Date());
```
Comment (lchanmann) [inline src/cronjobs]: Please use LocalDateTime if you can.

File: src/cronjobs (hunk 2 above)
Comment (KolombetM) [reply src/cronjobs]: I can't do that if I want to remove file from file system. 
Do I need to? @lchanmann

Comment (lchanmann) [reply src/cronjobs]: You can add TempFileDAO decorator to handle removing the file from file system then can just call removeAll() in the cron job.

File: src/cronjobs [hunk 5]
```diff
@@ -116,3 +116,51 @@ p({
     }
   """
 })
+
+p({
+  "class":"foam.nanos.cron.Cron",
+  "id":"CleanupTempFileCron",
+  "description":"Cron to clean up files which exist more than 4 hours",
+  schedule:{
+    class:"foam.nanos.cron.IntervalSchedule",
```
Comment (lchanmann) [inline src/cronjobs]: please use CronSchedule.

File: src/cronjobs [hunk 6]
```diff
@@ -116,3 +116,51 @@ p({
...
+
+    TempFile tempFile;
+
+    for ( int i = 0 ; i < tempFiles.size() ; i++ ) {
+      tempFile = (TempFile) tempFiles.get(i);
+       if ( forHoursBefore.after(tFile.getCreated()) ) {
+        tempFiles.get(i).getFile_().delete();
+        tempFileDAO.remove(tempFile);
+      }
+    }
```
Comment (lchanmann) [inline src/cronjobs]: Can move the logic for removing file from the file system in a decorator then can just call remove all on the filtered DAO.
```suggestion
    tempFileDAO
      .where(MLang.LT(TempFile.CREATED, threshold))
      .removeAll()
```

File: src/cronjobs [hunk 7]
```diff
@@ -116,3 +116,51 @@ p({
...
+  """,
+  "enabled":true
```
Comment (lchanmann) [inline src/cronjobs]:
```suggestion
  "enabled":false
```

File: src/cronjobs (hunk 2 above)
Comment (jlhughes) [reply src/cronjobs]: I'm fine with walking the list returned by select, but the where/select can handle the conditional testing.

**Review by lchanmann — COMMENTED** [review]
Please change the base branch to NP-5336/RepoDocument. @Mykola97

**Comment by jlhughes** [comment]
This is fine pulling into dev and then dev into your branch.  I prefer we review on development.

**Review by jlhughes — APPROVED** [review]

**Comment by lchanmann** [comment]
TempFile model is not in dev branch yet.

**Review by jlhughes — APPROVED** [review]

File: src/foam/nanos/fs/TempFileDAO.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  extends: 'foam.dao.ProxyDAO',
+
+  documentation: 'tempFileDAO decorator to perform remove operation',
+
+  methods: [
+    {
+      name: 'remove_',
+      documentation: 'TempFile stores actual file on fileSystem, so we need to explicitly remove it',
+      javaCode: `
+        ((TempFile) obj).getFile_().delete();
```
Comment (lchanmann) [inline src/foam/nanos/fs/TempFileDAO.js]: A small enhancement: can check if the `file_` is set then delete thus avoid invoking the factory to create the file just to delete it right immediately.

**Pull request merged**